File: app/src/main/java/org/lineageos/recorder/ListActivity.java
/*
 * Copyright (C) 2021 The LineageOS Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.lineageos.recorder;

import android.net.Uri;
import android.os.Bundle;
import android.text.Editable;
import android.view.ActionMode;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import org.lineageos.recorder.list.ListActionModeCallback;
import org.lineageos.recorder.list.RecordingData;
import org.lineageos.recorder.list.RecordingListCallbacks;
import org.lineageos.recorder.list.RecordingsAdapter;
import org.lineageos.recorder.task.DeleteAllRecordingsTask;
import org.lineageos.recorder.task.DeleteRecordingTask;
import org.lineageos.recorder.task.GetRecordingsTask;
import org.lineageos.recorder.task.RenameRecordingTask;
import org.lineageos.recorder.task.TaskExecutor;
import org.lineageos.recorder.utils.RecordIntentHelper;
import org.lineageos.recorder.utils.Utils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public class ListActivity extends AppCompatActivity implements RecordingListCallbacks {
    private static final String TYPE_AUDIO = "audio/*";

    private RecordingsAdapter mAdapter;

    private Toolbar mToolbar;
    private RecyclerView mListView;
    private ProgressBar mProgressBar;
    private TextView mEmptyText;

    private ActionMode mActionMode;

    private TaskExecutor mTaskExecutor;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        mTaskExecutor = new TaskExecutor();
        getLifecycle().addObserver(mTaskExecutor);

        setContentView(R.layout.activity_list);

        final CoordinatorLayout coordinatorLayout = findViewById(R.id.coordinator);
        mToolbar = findViewById(R.id.toolbar);
        mListView = findViewById(R.id.list_view);
        mProgressBar = findViewById(R.id.list_loading);
        mEmptyText = findViewById(R.id.list_empty);

        setSupportActionBar(mToolbar);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayShowHomeEnabled(true);
            actionBar.setDisplayHomeAsUpEnabled(true);
        }

        mAdapter = new RecordingsAdapter(this);
        mAdapter.registerAdapterDataObserver(new RecyclerView.AdapterDataObserver() {
            @Override
            public void onItemRangeRemoved(int positionStart, int itemCount) {
                super.onItemRangeRemoved(positionStart, itemCount);
                if (mAdapter.getItemCount() == 0) {
                    changeEmptyView(true);
                    endSelectionMode();
                }
            }
        });
        mListView.setLayoutManager(new LinearLayoutManager(this));
        mListView.setAdapter(mAdapter);

        loadRecordings();

        Utils.setFullScreen(getWindow(), coordinatorLayout);
        Utils.setVerticalInsets(mListView);
    }

    @Override
    public void onPlay(@NonNull Uri uri) {
        startActivity(RecordIntentHelper.getOpenIntent(uri, TYPE_AUDIO));
    }

    @Override
    public void onShare(@NonNull Uri uri) {
        startActivity(RecordIntentHelper.getShareIntent(uri, TYPE_AUDIO));
    }

    @Override
    public void onDelete(int index, @NonNull Uri uri) {
        new AlertDialog.Builder(this)
                .setTitle(R.string.delete_title)
                .setMessage(getString(R.string.delete_recording_message))
                .setPositiveButton(R.string.delete, (d, which) -> mTaskExecutor.runTask(
                        new DeleteRecordingTask(getContentResolver(), uri), () -> {
                            mAdapter.onDelete(index);
                            Utils.cancelShareNotification(this);
                        }))
                .setNegativeButton(R.string.cancel, null)
                .show();
    }

    @Override
    public void onRename(int index, @NonNull Uri uri, @NonNull String currentTitle) {
        final LayoutInflater inflater = getSystemService(LayoutInflater.class);
        final View view = inflater.inflate(R.layout.dialog_content_rename, null);
        EditText editText = view.findViewById(R.id.name);
        editText.setText(currentTitle);
        editText.requestFocus();
        Utils.showKeyboard(this);

        new AlertDialog.Builder(this)
                .setTitle(R.string.list_edit_title)
                .setView(view)
                .setPositiveButton(R.string.list_edit_confirm, (d, which) -> {
                    Editable editable = editText.getText();
                    if (editable == null || editable.length() == 0) {
                        return;
                    }

                    String newTitle = editable.toString();
                    if (!newTitle.equals(currentTitle)) {
                        renameRecording(uri, newTitle, index);
                    }
                    Utils.closeKeyboard(this);
                })
                .setNegativeButton(R.string.cancel, (d, which) -> Utils.closeKeyboard(this))
                .show();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        super.onCreateOptionsMenu(menu);
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_list, menu);
        return true;
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        MenuItem deleteAllItem = menu.findItem(R.id.action_delete_all);
        boolean hasItems = mAdapter.getItemCount() > 0;
        deleteAllItem.setEnabled(hasItems);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.action_delete_all) {
            promptDeleteAllRecordings();
            return true;
        } else {
            return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void startSelectionMode() {
        // Clear previous (should do nothing), but be sure
        endSelectionMode();
        // Start action mode
        mActionMode = mToolbar.startActionMode(new ListActionModeCallback(
                this::deleteSelectedRecordings,
                this::shareSelectedRecordings));
        mAdapter.enterSelectionMode();
    }

    @Override
    public void endSelectionMode() {
        if (mActionMode != null) {
            mActionMode.finish();
            mActionMode = null;
        }
        mAdapter.exitSelectionMode();
    }

    @Override
    public void onActionModeFinished(@NonNull ActionMode mode) {
        super.onActionModeFinished(mode);
        endSelectionMode();
    }

    private void loadRecordings() {
        mTaskExecutor.runTask(new GetRecordingsTask(getContentResolver()), list -> {
            mProgressBar.setVisibility(View.GONE);
            mAdapter.setData(list);
            changeEmptyView(list.isEmpty());
        });
    }

    private void renameRecording(@NonNull Uri uri, @NonNull String newTitle, int index) {
        mTaskExecutor.runTask(new RenameRecordingTask(getContentResolver(), uri, newTitle),
                success -> {
                    if (success) {
                        mAdapter.onRename(index, newTitle);
                    }
                });
    }

    private void deleteRecording(@NonNull RecordingData item) {
        mTaskExecutor.runTask(new DeleteRecordingTask(getContentResolver(), item.getUri()),
                () -> {
                    mAdapter.onDelete(item);
                    Utils.cancelShareNotification(this);
                });
    }

    private void deleteAllRecordings() {
        final List<Uri> uris = mAdapter.getData().stream()
                .map(RecordingData::getUri)
                .collect(Collectors.toList());
        mTaskExecutor.runTask(new DeleteAllRecordingsTask(getContentResolver(), uris), () -> {
            mAdapter.setData(Collections.emptyList());
            changeEmptyView(true);
        });
    }

    private void changeEmptyView(boolean isEmpty) {
        mEmptyText.setVisibility(isEmpty ? View.VISIBLE : View.GONE);
        mListView.setVisibility(isEmpty ? View.GONE : View.VISIBLE);
    }

    private void shareSelectedRecordings() {
        final List<RecordingData> selectedItems = mAdapter.getSelected();
        if (selectedItems.isEmpty()) {
            return;
        }

        final ArrayList<Uri> uris = selectedItems.stream()
                .map(RecordingData::getUri)
                .collect(Collectors.toCollection(ArrayList::new));
        startActivity(RecordIntentHelper.getShareIntents(uris, TYPE_AUDIO));
    }

    private void deleteSelectedRecordings() {
        final List<RecordingData> selectedItems = mAdapter.getSelected();
        if (selectedItems.isEmpty()) {
            return;
        }

        new AlertDialog.Builder(this)
                .setTitle(R.string.delete_selected_title)
                .setMessage(getString(R.string.delete_selected_message))
                .setPositiveButton(R.string.delete, (dialog, which) -> {
                    selectedItems.forEach(this::deleteRecording);
                    Utils.cancelShareNotification(this);
                })
                .setNegativeButton(R.string.cancel, null)
                .show();
    }

    private void promptDeleteAllRecordings() {
        if (mAdapter.getItemCount() == 0) {
            return;
        }

        new AlertDialog.Builder(this)
                .setTitle(R.string.delete_all_title)
                .setMessage(getString(R.string.delete_all_message))
                .setPositiveButton(R.string.delete, (dialog, which) -> {
                    deleteAllRecordings();
                    Utils.cancelShareNotification(this);
                })
                .setNegativeButton(R.string.cancel, null)
                .show();
    }
}

File: app/src/main/java/org/lineageos/recorder/ListActivity.kt
/*
 * SPDX-FileCopyrightText: 2021-2024 The LineageOS Project
 * SPDX-License-Identifier: Apache-2.0
 */

package org.lineageos.recorder

import android.content.DialogInterface
import android.os.Bundle
import android.view.ActionMode
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.EditText
import android.widget.FrameLayout
import android.widget.ProgressBar
import android.widget.TextView
import androidx.activity.viewModels
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.view.ViewCompat
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.isVisible
import androidx.core.view.updatePadding
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.Observer
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.selection.SelectionTracker
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import org.lineageos.recorder.ext.getCurrentSelection
import org.lineageos.recorder.list.ListActionModeCallback
import org.lineageos.recorder.list.RecordingItemCallbacks
import org.lineageos.recorder.models.Recording
import org.lineageos.recorder.list.RecordingsAdapter
import org.lineageos.recorder.utils.RecordIntentHelper
import org.lineageos.recorder.viewmodels.RecordingsViewModel
import kotlin.reflect.cast

class ListActivity : AppCompatActivity() {
    // View models
    private val model: RecordingsViewModel by viewModels()

    // Views
    private val contentView by lazy { findViewById<View>(android.R.id.content) }
    private val listEmptyTextView by lazy { findViewById<TextView>(R.id.listEmptyTextView) }
    private val listLoadingProgressBar by lazy { findViewById<ProgressBar>(R.id.listLoadingProgressBar) }
    private val listRecyclerView by lazy { findViewById<RecyclerView>(R.id.listRecyclerView) }
    private val toolbar by lazy { findViewById<Toolbar>(R.id.toolbar) }

    // Adapters
    private val recordingItemCallbacks = object : RecordingItemCallbacks {
        override fun onPlay(recording: Recording) {
            this@ListActivity.onPlay(recording)
        }

        override fun onShare(recording: Recording) {
            this@ListActivity.onShare(recording)
        }

        override fun onDelete(recording: Recording) {
            this@ListActivity.onDelete(recording)
        }

        override fun onRename(recording: Recording) {
            this@ListActivity.onRename(recording)
        }
    }
    private val recordingsAdapter by lazy { RecordingsAdapter(recordingItemCallbacks) }

    // Selection
    private var selectionTracker: SelectionTracker<Recording>? = null

    private val selectionTrackerObserver =
        object : SelectionTracker.SelectionObserver<Recording>() {
            override fun onSelectionChanged() {
                super.onSelectionChanged()

                updateSelection()
            }

            override fun onSelectionRefresh() {
                super.onSelectionRefresh()

                updateSelection()
            }

            override fun onSelectionRestored() {
                super.onSelectionRestored()

                updateSelection()
            }
        }

    private var actionMode: ActionMode? = null

    private val actionModeCallback = object : ActionMode.Callback {
        override fun onCreateActionMode(mode: ActionMode?, menu: Menu?): Boolean {
            requireActivity().menuInflater.inflate(
                when (bucketId) {
                    MediaStoreBuckets.MEDIA_STORE_BUCKET_TRASH.id -> R.menu.album_action_bar_trash
                    else -> R.menu.album_action_bar
                },
                menu
            )
            return true
        }

        override fun onPrepareActionMode(mode: ActionMode?, menu: Menu?) = false

        override fun onActionItemClicked(mode: ActionMode?, item: MenuItem?) =
            selectionTracker?.selection?.toList()?.toTypedArray()?.takeUnless {
                it.isEmpty()
            }?.let { selection ->
                when (item?.itemId) {
                    R.id.deleteForever -> {
                        MediaDialogsUtils.openDeleteForeverDialog(requireContext(), *selection) {
                            deleteForeverContract.launch(
                                requireContext().contentResolver.createDeleteRequest(
                                    *it.map { media ->
                                        media.uri
                                    }.toTypedArray()
                                )
                            )
                        }

                        true
                    }

                    R.id.share -> {
                        requireActivity().startActivity(buildShareIntent(*selection))

                        true
                    }

                    else -> false
                }
            } ?: false

        override fun onDestroyActionMode(mode: ActionMode?) {
            selectionTracker?.clearSelection()
        }
    }

    private val inSelectionModeObserver = Observer { inSelectionMode: Boolean ->
        if (inSelectionMode) {
            startSelectionMode()
        } else {
            endSelectionMode()
        }
    }

    private var lastProcessedSelection: Array<out Recording>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_list)

        // Setup edge-to-edge
        WindowCompat.setDecorFitsSystemWindows(window, false)

        setSupportActionBar(toolbar)
        supportActionBar?.let {
            it.setDisplayShowHomeEnabled(true)
            it.setDisplayHomeAsUpEnabled(true)
        }

        listRecyclerView.layoutManager = LinearLayoutManager(this)
        listRecyclerView.adapter = recordingsAdapter

        ViewCompat.setOnApplyWindowInsetsListener(contentView) { _, windowInsets ->
            val insets = windowInsets.getInsets(WindowInsetsCompat.Type.systemBars())

            listRecyclerView.updatePadding(
                bottom = insets.bottom,
                left = insets.left,
                right = insets.right,
            )

            windowInsets
        }

        lifecycleScope.launch {
            lifecycle.repeatOnLifecycle(Lifecycle.State.STARTED) {
                model.recordings.collectLatest {
                    recordingsAdapter.submitList(it)

                    listLoadingProgressBar.isVisible = false

                    val isEmpty = it.isEmpty()
                    changeEmptyView(isEmpty)
                    if (isEmpty) {
                        endSelectionMode()
                    }
                }
            }
        }
    }

    fun onPlay(recording: Recording) {
        startActivity(RecordIntentHelper.getOpenIntent(recording.uri, TYPE_AUDIO))
    }

    fun onShare(recording: Recording) {
        startActivity(RecordIntentHelper.getShareIntent(recording.uri, TYPE_AUDIO))
    }

    fun onDelete(recording: Recording) {
        MaterialAlertDialogBuilder(this)
            .setTitle(R.string.delete_title)
            .setMessage(R.string.delete_recording_message)
            .setPositiveButton(R.string.delete) { _: DialogInterface?, _: Int ->
                lifecycleScope.launch {
                    model.deleteRecordings(recording)
                }
            }
            .setNegativeButton(R.string.cancel, null)
            .show()
    }

    fun onRename(recording: Recording) {
        lateinit var alertDialog: AlertDialog
        lateinit var editText: EditText

        val onConfirm = {
            editText.text?.takeIf { it.isNotEmpty() }?.let { editable ->
                lifecycleScope.launch {
                    model.renameRecording(recording, editable.toString())
                }

                true
            } ?: false
        }

        val view = FrameLayout::class.cast(
            layoutInflater.inflate(
                R.layout.dialog_content_rename,
                null,
                false
            )
        )
        editText = view.findViewById<EditText>(R.id.nameEditText).apply {
            setText(recording.title)
            setSelection(0, recording.title.length)
            setOnEditorActionListener { _, actionId, _ ->
                when (actionId) {
                    EditorInfo.IME_ACTION_UNSPECIFIED,
                    EditorInfo.IME_ACTION_DONE -> {
                        onConfirm().also {
                            if (it) {
                                alertDialog.dismiss()
                            }
                        }
                    }

                    else -> false
                }
            }
        }

        alertDialog = MaterialAlertDialogBuilder(this)
            .setTitle(R.string.list_edit_title)
            .setView(view)
            .setPositiveButton(R.string.list_edit_confirm) { _, _ -> onConfirm() }
            .setNegativeButton(R.string.cancel, null)
            .show()
            .also {
                editText.requestFocus()
            }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        super.onCreateOptionsMenu(menu)
        val inflater = menuInflater
        inflater.inflate(R.menu.menu_list, menu)
        return true
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        val deleteAllItem = menu.findItem(R.id.action_delete_all)
        val hasItems = recordingsAdapter.itemCount > 0
        deleteAllItem.setEnabled(hasItems)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem) = when (item.itemId) {
        R.id.action_delete_all -> {
            promptDeleteAllRecordings()
            true
        }

        else -> false
    }

    fun startSelectionMode() {
        // Clear previous (should do nothing), but be sure
        endSelectionMode()
        // Start action mode
        actionMode = toolbar.startActionMode(
            ListActionModeCallback({
                deleteSelectedRecordings()
            }) { shareSelectedRecordings() }
        )
        // TODO
        //adapter.enterSelectionMode()
    }

    fun endSelectionMode() {
        actionMode?.finish()
        actionMode = null

        // TODO
        //adapter.exitSelectionMode()
    }

    override fun onActionModeFinished(mode: ActionMode) {
        super.onActionModeFinished(mode)
        endSelectionMode()
    }

    private fun changeEmptyView(isEmpty: Boolean) {
        listEmptyTextView.isVisible = isEmpty
        listRecyclerView.isVisible = !isEmpty
    }

    private fun shareSelectedRecordings() {
        selectionTracker?.getCurrentSelection()
            ?.toList()
            ?.takeIf { it.isNotEmpty() }
            ?.let { selectedItems ->
                val uris = selectedItems
                    .map { it.uri }

                startActivity(RecordIntentHelper.getShareIntents(uris, TYPE_AUDIO))
            }
    }

    private fun deleteSelectedRecordings() {
        selectionTracker?.getCurrentSelection()
            ?.toList()
            ?.takeIf { it.isNotEmpty() }
            ?.let { selectedItems ->
                MaterialAlertDialogBuilder(this)
                    .setTitle(R.string.delete_selected_title)
                    .setMessage(getString(R.string.delete_selected_message))
                    .setPositiveButton(R.string.delete) { _: DialogInterface?, _: Int ->
                        lifecycleScope.launch {
                            model.deleteRecordings(*selectedItems.toTypedArray())
                        }
                    }
                    .setNegativeButton(R.string.cancel, null)
                    .show()
            }
    }

    private fun promptDeleteAllRecordings() {
        if (recordingsAdapter.itemCount == 0) {
            return
        }

        MaterialAlertDialogBuilder(this)
            .setTitle(R.string.delete_all_title)
            .setMessage(getString(R.string.delete_all_message))
            .setPositiveButton(R.string.delete) { _: DialogInterface?, _: Int ->
                lifecycleScope.launch {
                    model.deleteRecordings(*recordingsAdapter.currentList.toTypedArray())
                }
            }
            .setNegativeButton(R.string.cancel, null)
            .show()
    }

    companion object {
        private const val TYPE_AUDIO = "audio/*"
    }
}
